draft-ietf-masque-connect-udp-listen.md

```diff
@@ -164,6 +166,39 @@ and therefore is to be ignored). This document does not define any parameters
 for the Connect-UDP-Bind header field value, but future documents might define
 parameters. Receivers MUST ignore unknown parameters.
 
+
+# The Proxy-Public-Address Response Header Field {#response}
+
+Upon accepting the request, the proxy MUST select at least one public IP
+address to bind. The proxy MAY also select a second address from a different address
+family. For each selected address, it MUST select an open port to
+bind to this request. From then and until the tunnel is closed, the proxy
+SHALL send packets received on these IP-port tuples to the client. The proxy
+then MUST communicate the selected addresses and ports to the client using
+the "Proxy-Public-Address" header. The format of that header is defined below
+using IPv4address, IPv6address and port from {{Section 3.2 of !URI=RFC3986}}.
+
+The proxy MUST then respond with the allocated IP and port pairings using the
+Proxy-Public-Address response header which is defined as a List of Strings.
+The list contains a maximum of one IPv4 address and port, and one IPv6 address
+and port each. IP-literal, IPv4address and Port are defined in
+{{Section 3 of !URI=RFC3986}}
+
+~~~ ascii-art
+proxy-public-address = List of IP:Port pairings
+IP:Port pairing = IP-literal:Port / IPv4address:Port
+~~~
+{: #target-format title="Proxy Address Format"}
```

Comment:
Speaking of which we should reference ABNF=RFC5234

Reply:
> Thinking through this some more, maybe it's best to not restrict to two? I don't remember what the motivation was for the restriction

The reason was up to one for IPv4 and one for IPv6

Reply:
I thought so too but github isn't letting me reply to that comment for some weird reason

Reply:
What I meant was: I don't remember why we restricted to one IPv4 and one IPv6

Reply:
I mean I guess you could have more than one of each (to work with different subnets?). But then we need a way for the client to know which IP:Port would be used for what target

Reply:
The client doesn't need to know.  Even the proxy may not know!  On many operating systems, the choice of outgoing interface is controlled by the routing table and is not straightforward for application software to control.

Reply:
Would having multiple addresses for the same family be counterproductive to this extensions since the client can no longer guarantee the IP:Port at which it will be sending/receiving.

Reply:
For peer-to-peer use cases like WebRTC, it's normal to have multiple addresses of the same family.  Each address becomes an "ICE candidate", and the ICE procedure tries all pairs of addresses (in some prioritized order) to find a working connection (accounting for interference by NATs, etc.).

For client-server use cases like HTTP/3, the situation is perhaps more interesting.  If the server is trying to use this protocol, it likely does need a single consistent address for each connection.  It also needs other guarantees beyond the proxy itself, such as requiring that there are no firewalls or NATs preventing inbound connections to the ports it has opened.

My recommendation would be to remove any restriction on the number of addresses within the protocol, but note that proxies intended for client-server use should select at most one address of each family.